litex/soc/cores/spiopi.py

```diff
@@ -0,0 +1,1063 @@
+        self.specials += [
+            Instance("BUFR", i_I=pads.dqs, o_O=dqs_iobuf),
+        ]
+
+
+        # DQ connections -------------------------------------------------------------------------
+        # PHY API
+        self.do = Signal(16) # OPI data to SPI
+        self.di = Signal(16) # OPI data from SPI
+        self.tx = Signal() # when asserted OPI is transmitting data to SPI, otherwise, receiving
+
+        self.mosi = Signal() # SPI data to SPI
+        self.miso = Signal() # SPI data from SPI
+
+        # Delay programming API
+        self.delay_config = CSRStorage(fields=[
+            CSRField("d", size=5, description="Delay amount; each increment is 78ps", reset=31),
+            CSRField("load", size=1, description="Force delay taps to delay_d"),
+        ])
+        self.delay_status = CSRStatus(fields=[
+            CSRField("q", size=5, description="Readback of current delay amount, useful if inc/ce is used to set"),
+        ])
+        self.delay_update = Signal()
+        self.hw_delay_load = Signal()
+        self.sync += self.delay_update.eq(self.hw_delay_load | self.delay_config.fields.load)
+
+        # Break system API into rising/falling edge samples
+        do_rise = Signal(8) # data output presented on the rising edge
+        do_fall = Signal(8) # data output presented on the falling edge
+        self.comb += [do_rise.eq(self.do[8:]), do_fall.eq(self.do[:8])]
+
+        di_rise = Signal(8)
+        di_fall = Signal(8)
+        self.comb += self.di.eq(Cat(di_fall, di_rise))
+
+        # OPI DDR registers
+        dq = TSTriple(7) # dq[0] is special because it is also MOSI
+        dq_delayed = Signal(8)
+        self.specials += dq.get_tristate(pads.dq[1:])
+        for i in range(1, 8):
+            self.specials += Instance("ODDR",
+                p_DDR_CLK_EDGE="SAME_EDGE",
+                i_C=ClockSignal(), i_R=ResetSignal(), i_S=0, i_CE=1,
+                i_D1=do_rise[i], i_D2=do_fall[i], o_Q=dq.o[i-1],
+            )
+            if sim == False:
+                if i == 1: # only wire up o_CNTVALUEOUT for one instance
+                    self.specials += Instance("IDELAYE2",
+                             p_DELAY_SRC="IDATAIN", p_SIGNAL_PATTERN="DATA",
+                             p_CINVCTRL_SEL="FALSE", p_HIGH_PERFORMANCE_MODE="FALSE", p_REFCLK_FREQUENCY=200.0,
+                             p_PIPE_SEL="FALSE", p_IDELAY_VALUE=dq_delay_taps, p_IDELAY_TYPE=delay_type,
+
+                             i_C=ClockSignal(), i_CINVCTRL=0, i_REGRST=0, i_LDPIPEEN=0, i_INC=0, i_CE=0,
+                             i_LD=self.delay_update,
+                             i_CNTVALUEIN=self.delay_config.fields.d, o_CNTVALUEOUT=self.delay_status.fields.q,
+                             i_IDATAIN=dq.i[i-1], o_DATAOUT=dq_delayed[i],
+                    ),
+                else: # don't wire up o_CNTVALUEOUT for others
+                    self.specials += Instance("IDELAYE2",
```

Comment:
I would suggest wrapping right after the first brace so everything isn't indented so much.
(Same with the state machine below).